File: src/python/core/src/dicebear/utils/initials.py
"""Derive display initials from a seed string."""

from __future__ import annotations

import re
import unicodedata

# Backtick, acute accent, apostrophe, and modifier letter apostrophe — stripped
# so names like "O'Brien" yield "OB" rather than "OB" split on the quote.
_QUOTES = re.compile(r"[`´'ʼ]")
_AT_TAIL = re.compile(r"@.*")


class Initials:
    """Derive display initials from a seed string."""

    @staticmethod
    def from_seed(seed: str, discard_at_symbol: bool = True) -> str:
        """Return one or two uppercase initials for the given seed.

        By default strips ``@...`` so email addresses yield a single initial
        instead of being treated as two words.
        """
        text = _AT_TAIL.sub("", seed) if discard_at_symbol else seed
        text = _QUOTES.sub("", text)

        words = _words(text)

        if len(words) == 0:
            return Initials.from_seed(seed, False) if discard_at_symbol else ""

        if len(words) == 1:
            return _clusters(words[0], 2).upper()

        first = _clusters(words[0], 1)
        last = _clusters(words[-1], 1)

        if first == "" or last == "":
            return ""

        return (first + last).upper()


def _is_letter(char: str) -> bool:
    """Whether ``char`` is a Unicode letter (PCRE ``\\p{L}``)."""
    return unicodedata.category(char)[0] == "L"


def _is_mark(char: str) -> bool:
    """Whether ``char`` is a Unicode combining mark (PCRE ``\\p{M}``)."""
    return unicodedata.category(char)[0] == "M"


def _words(text: str) -> list[str]:
    """Split into maximal runs that start with a letter and continue with
    letters or marks — the matches of ``(\\p{L}[\\p{L}\\p{M}]*)``.
    """
    words: list[str] = []
    current: list[str] = []

    for char in text:
        if current:
            if _is_letter(char) or _is_mark(char):
                current.append(char)
            else:
                words.append("".join(current))
                current = []
                if _is_letter(char):
                    current.append(char)
        elif _is_letter(char):
            current.append(char)

    if current:
        words.append("".join(current))

    return words


def _clusters(word: str, limit: int) -> str:
    """Return up to ``limit`` leading letter-plus-marks clusters of ``word``.

    Mirrors ``(?:\\p{L}\\p{M}*){limit}`` anchored at the start.
    """
    result: list[str] = []
    count = 0
    i = 0

    while i < len(word) and count < limit:
        if not _is_letter(word[i]):
            break

        result.append(word[i])
        i += 1

        while i < len(word) and _is_mark(word[i]):
            result.append(word[i])
            i += 1

        count += 1

    return "".join(result)

File: src/go/core/internal/initials/initials.go
// Package initials derives display initials from a seed string.
//
// Words are split by Unicode letter/mark classes (\p{L}, \p{M}) so the result
// matches every other port for accented and non-Latin input.
// See https://www.regular-expressions.info/unicode.html
package initials

import (
	"regexp"
	"strings"

	"golang.org/x/text/cases"
	"golang.org/x/text/language"
)

// upper applies full Unicode case mapping (ß→SS, ﬁ→FI, …), matching JS
// toUpperCase and Rust to_uppercase. Go's strings.ToUpper only does simple 1:1
// mapping, which would diverge for special-casing letters. A fresh Caser is
// built per call because cases.Caser is not safe for concurrent use.
func upper(s string) string {
	return cases.Upper(language.Und).String(s)
}

var (
	// Apostrophes and accents that should not break a word (e.g. l'eau, d´or).
	reQuotes = regexp.MustCompile("[`´'ʼ]")
	// A word: a letter followed by any letters or combining marks.
	reWord = regexp.MustCompile(`\p{L}[\p{L}\p{M}]*`)
	// The first one or two letter+marks units of a word.
	reTwo = regexp.MustCompile(`^(?:\p{L}\p{M}*){1,2}`)
	// The first single letter+marks unit of a word.
	reOne = regexp.MustCompile(`^(?:\p{L}\p{M}*)`)
)

// FromSeed returns one or two uppercase initials for seed, discarding an "@…"
// suffix so email addresses yield a single initial.
func FromSeed(seed string) string {
	return fromSeedWith(seed, true)
}

// fromSeedWith is FromSeed with control over whether an "@…" suffix is stripped
// before deriving initials.
func fromSeedWith(seed string, discardAtSymbol bool) string {
	input := seed

	if discardAtSymbol {
		// Strip the whole @ suffix (including any line terminators), mirroring
		// the dotall `@.*` strip in every port. Only the first @ matters since
		// `.*` reaches the end.
		if i := strings.IndexByte(input, '@'); i >= 0 {
			input = input[:i]
		}
	}

	input = reQuotes.ReplaceAllString(input, "")

	words := reWord.FindAllString(input, -1)

	if len(words) == 0 {
		if discardAtSymbol {
			return fromSeedWith(seed, false)
		}
		return ""
	}

	if len(words) == 1 {
		if m := reTwo.FindString(words[0]); m != "" {
			return upper(m)
		}
		return ""
	}

	first := reOne.FindString(words[0])
	last := reOne.FindString(words[len(words)-1])

	if first == "" || last == "" {
		return ""
	}

	return upper(first + last)
}
